### PR TITLE
Added a flow for setting volume profiling for all the volumes

### PR DESCRIPTION
**Pull request by shtripat**
tendrl-bug-id: Tendrl/gluster-integration#592
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@nthomas-redhat please review. Still needs to be verified.

**Comment by shtripat** [comment]
Tested the flow through a test program and works as expected. Sample input for the flow is `Cluster.volume_profiling_flag=enable` or `Cluster.volume_profiling_flag=disable`.
Will send a PR  for adding API for the same.

**Comment by shtripat** [comment]
This PR also fixes the volume level profiling enable and disable. The atoms now verify the changes against the underlying volume in the cluster and accordingly set the value of `Volume.profiling_enabled` value as `yes` or `no`. The jobs would be very short lived and sometimes UI might not be able to show `Pending` state for the volume profiling.

File: tendrl/gluster_integration/objects/volume/atoms/stop_profiling/__init__.py [hunk 1]
```diff
@@ -27,12 +27,24 @@ def run(self):
...
             )
+        loop_count = 0
         while True:
-            volume = NS.gluster.objects.Volume(
-                vol_id=vol_id
-            ).load()
-            if volume.profiling_enabled == "no":
+            if loop_count >= 24:
+                raise AtomExecutionFailedError(
+                    "Volume profiling not yet marked disabled "
```
Comment (r0h4n) [inline tendrl/gluster_integration/objects/volume/atoms/stop_profiling/__init__.py]: "Could not disable profiling for volume: %s under cluster: %s, Timed out"

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 2]
```diff
@@ -340,62 +341,71 @@ def _enable_disable_volume_profiling(self):
...
-            elif profiling_enabled_count == len(volumes):
+        # Enable / disable based on cluster flag volume_profiling_flag
+        # should be done only once while first sync. Later the volume
+        # level volume_profiling_state should be set based on individual
+        # volume level values
+        first_sync_done = etcd_utils.read(
+            "/clusters/%s/nodes/%s/NodeContext/first_sync_done" %
+            (NS.tendrl_context.integration_id,
+             NS.node_context.node_id)
+        ).value
```
Comment (r0h4n) [inline tendrl/gluster_integration/sds_sync/__init__.py]: Why direct read, use NS.tendrl.objects.ClusterNodeContext()

File: tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        volumes = NS.gluster.objects.Volume().load_all() or []
+        failed_vols = []
+        for volume in volumes:
+            out, err, rc = cmd_utils.Command(
+                "gluster volume profile %s %s" %
+                (volume.name, VOL_PROFILE_ACTIONS[action])
+            ).run()
+            if err or rc != 0:
+                logger.log(
+                    "INFO",
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py]: ok

File: tendrl/gluster_integration/objects/volume/atoms/start_profiling/__init__.py [hunk 4]
```diff
@@ -27,12 +27,25 @@ def run(self):
...
             )
+        loop_count = 0
         while True:
-            volume = NS.gluster.objects.Volume(
-                vol_id=vol_id
-            ).load()
-            if volume.profiling_enabled == "yes":
+            if loop_count >= 24:
+                raise AtomExecutionFailedError(
+                    "Volume profiling not yet marked enabled "
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/gluster_integration/objects/volume/atoms/start_profiling/__init__.py]: ok

File: tendrl/gluster_integration/objects/volume/atoms/stop_profiling/__init__.py (hunk 1 above)
Comment (shtripat) [reply tendrl/gluster_integration/objects/volume/atoms/stop_profiling/__init__.py]: ok

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 2 above)
Comment (shtripat) [reply tendrl/gluster_integration/sds_sync/__init__.py]: I remember I faced some issues while reading `first_sync_done` field. And as a quick solution I did this. Will check that.

File: tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        super(EnableDisableVolumeProfiling, self).__init__(
+            *args,
+            **kwargs
+        )
+
+    def run(self):
+        action = self.parameters["Cluster.volume_profiling_flag"]
+        if action not in VOL_PROFILE_ACTIONS.keys():
+            raise FlowExecutionFailedError(
+                "Invalid value of Cluster.volume_profiling_flag "
```
Comment (r0h4n) [inline tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py]: Please mention in log what values are valid i.e "enable" | "disable"

File: tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    NS.tendrl_context.integration_id
+                )
+            )
+
+        _cluster = NS.tendrl.objects.Cluster(
+            integration_id=NS.tendrl_context.integration_id
+        ).load()
+        _lock_details = {
+            'node_id': NS.node_context.node_id,
+            'fqdn': NS.node_context.fqdn,
```
Comment (r0h4n) [inline tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py]: fqdn not required

File: tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py (hunk 5 above)
Comment (shtripat) [reply tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py]: Done

File: tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py (hunk 6 above)
Comment (shtripat) [reply tendrl/gluster_integration/flows/enable_disable_volume_profiling/__init__.py]: Done. Will remove that

**Comment by shtripat** [comment]
Yes its verified from UI as well

**Pull request merged**